aml_module/invoker.py

```diff
@@ -0,0 +1,167 @@
+import argparse
+from urllib.parse import unquote
+import pandas as pd
+import numpy as np
+from azureml.studio.core.logger import module_logger as logger
+from azureml.studio.core.io.data_frame_directory import load_data_frame_from_directory, save_data_frame_to_directory
+from azureml.studio.core.utils.column_selection import ColumnSelection
+from azureml.studio.internal.error_handler import error_handler
+import sr_detector
+from error_messages import *
+from constants import *
+
+PACKAGE_NAME = 'spectral_residual_anomaly_detection_module'
+VERSION = '0.0.1'
+
+
+def str2bool(v):
+    if isinstance(v, bool):
+        return v
+    if v.lower() in ('yes', 'true', 't', 'y', '1'):
+        return True
+    elif v.lower() in ('no', 'false', 'f', 'n', '0'):
+        return False
+    else:
+        raise argparse.ArgumentTypeError('Boolean value expected.')
+
+
+def is_timestamp_ascending(timestamps):
+    count = len(timestamps)
+
+    for i in range(count - 1):
+        if timestamps[i] > timestamps[i + 1]:
+            return -1
+        elif timestamps[i] == timestamps[i + 1]:
+            return -2
+    return 0
+
+
+@error_handler
+def invoke(input_path, detect_mode, timestamp_column, value_column, batch_size, threshold, sensitivity,
+            appendMode, compute_stats_in_visualization, output_path):
+    data_frame_directory = load_data_frame_from_directory(input_path)
+
+    logger.debug(f"Shape of loaded DataFrame: {data_frame_directory.data.shape}")
+
+    if data_frame_directory.data.shape[0] < MIN_POINTS:
+        raise Exception(NotEnoughPoints.format(MIN_POINTS))
```

Comment:
It would be better to use `ErrorMapping.verify_number_of_rows_greater_than_or_equal_to` to throw UserError in azureml.studio.internal.error

Reply:
Similar comment to other exceptions, all the errors inherited from the class `ModuleError` will be shown as UserError in Designer, which indicates that the input/param of the user is not correct.
Otherwise the errors will be raised as `LibraryError` which indicates that the exception is unexpected and need to be fixed.

Reply:
According to the latest update, you don't need to import `azureml.studio.internal.error`
You just need to import `UserError` from `azureml.studio.core.error`, and simply
`raise UserError(xxx)`.